File: patterns/03-data-messaging-patterns/04-publisher-subscriber/python/publisher_subscriber.py
class PatternError(Exception):
    def __init__(self, code):
        super().__init__(code)
        self.code = code

def evaluate(input_data):
    subscribers = input_data.get("subscribers", [])
    names = [subscriber["name"] for subscriber in subscribers]
    if len(names) != len(set(names)):
        raise PatternError("duplicate_subscriber")
    deliveries = []
    events = input_data.get("events", [])
    for event in events:
        for subscriber in subscribers:
            if event["topic"] not in subscriber.get("topics", []):
                continue
            outcome = subscriber.get("outcome", "success")
            if outcome not in ("success", "failure"):
                raise PatternError("unknown_delivery_outcome")
            deliveries.append({"event_id": event["id"], "subscriber": subscriber["name"], "status": "delivered" if outcome == "success" else "failed"})
    return {"published": len(events), "deliveries": deliveries}

File: patterns/03-data-messaging-patterns/04-publisher-subscriber/js/publisher_subscriber.mjs
export class PatternError extends Error { constructor(code) { super(code); this.code = code; } }

export function evaluate(input) {
    const subscribers = input.subscribers ?? [];
    const names = new Set();
    for (const subscriber of subscribers) {
        if (names.has(subscriber.name)) throw new PatternError("duplicate_subscriber");
        names.add(subscriber.name);
    }
    const events = input.events ?? [];
    const deliveries = [];
    for (const event of events) {
        for (const subscriber of subscribers) {
            if (!(subscriber.topics ?? []).includes(event.topic)) continue;
            const outcome = subscriber.outcome ?? "success";
            if (outcome !== "success" && outcome !== "failure") throw new PatternError("unknown_delivery_outcome");
            deliveries.push({event_id: event.id, subscriber: subscriber.name, status: outcome === "success" ? "delivered" : "failed"});
        }
    }
    return {published: events.length, deliveries};
}
